rust/src/types/stack.rs

```diff
@@ -1,19 +1,112 @@
+        let mut s = Self {
+            data: [const { MaybeUninit::uninit() }; 1024],
+            len,
+        };
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout
+        // With nightly rust this could be replaced by MaybeUninit::copy_from_slice
+        s.data[..len].copy_from_slice(unsafe {
+            std::mem::transmute::<&[u256], &[std::mem::MaybeUninit<u256>]>(inner)
+        });
+        s
+    }
+
+    pub fn as_slice(&self) -> &[u256] {
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout and self.data is initialized up to
+        // self.len
+        unsafe {
+            std::mem::transmute::<&[std::mem::MaybeUninit<u256>], &[u256]>(&self.data[..self.len])
+        }
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [u256] {
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout and self.data is initialized up to
+        // self.len
+        unsafe {
+            std::mem::transmute::<&mut [std::mem::MaybeUninit<u256>], &mut [u256]>(
+                &mut self.data[..self.len],
+            )
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub unsafe fn push(&mut self, value: u256) {
+        *self.data.get_unchecked_mut(self.len) = MaybeUninit::new(value);
+        self.len += 1;
+    }
+
+    pub fn pop_array<const N: usize>(&mut self) -> Option<[u256; N]> {
+        if self.len < N {
+            return None;
+        }
+        self.len -= N;
+        let start = self.data.as_ptr() as *const u256;
+        // SAFETY:
+        // This does not wrap and the whole range from start to start + self.len is valid.
+        let array_start = unsafe { start.add(self.len) };
+        // SAFETY:
+        // The invariant of this type is that the first `self.len` elements are initialized.
+        // `self.len` just got decremented by N, which means now that the first `self.len + N`
+        // elements are initialized. Therefore, it is safe to read N elements starting at index
+        // `self.len` as an array of length N and type u256.
+        unsafe { Some(std::ptr::read(array_start as *const [u256; N])) }
+    }
+}
+
+#[cfg(feature = "stack-array")]
+#[derive(Debug)]
+pub struct Stack(StackRaw);
+
+#[cfg(not(feature = "stack-array"))]
 #[derive(Debug)]
 pub struct Stack(Vec<u256>);
 
 impl Stack {
-    pub fn new(inner: Vec<u256>) -> Self {
-        Self(inner)
+    pub fn new(inner: &[u256]) -> Self {
+        #[cfg(feature = "stack-array")]
+        return Self(StackRaw::new(inner));
+        #[cfg(not(feature = "stack-array"))]
+        return Self(Vec::from(inner));
     }
 
-    pub fn into_inner(self) -> Vec<u256> {
-        self.0
+    pub fn as_slice(&self) -> &[u256] {
+        self.0.as_slice()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
     }
 
     pub fn push(&mut self, value: impl Into<u256>) -> Result<(), FailStatus> {
         self.check_overflow_on_push()?;
```

Comment:
if we move the overflow check inside of push (like pop does) 
we can do the overflow check only for the vector case.  and remove the unsafe block from this scope.

Reply:
See earlier comment. I don't like the idea of doing the check in different places.